OpenRA.Mods.Common/AI/States/NavyStates.cs

```diff
@@ -0,0 +1,194 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.AI
+{
+	abstract class NavyStateBase : StateBase
+	{
+		protected virtual bool ShouldFlee(Squad owner)
+		{
+			return base.ShouldFlee(owner, enemies => !AttackOrFleeFuzzy.Default.CanAttack(owner.Units, enemies));
+		}
+
+		protected Actor FindClosestEnemy(Squad owner)
+		{
+			if (!owner.IsValid)
+				return null;
+
+			// For ships, cheat and move towards enemy naval production, if any.
+			var t = owner.World.ActorsHavingTrait<Building>().Where(a
+				=> owner.Bot.Info.BuildingCommonNames.NavalProduction.Contains(a.Info.Name)
+				&& a.AppearsHostileTo(owner.Units.First())).FirstOrDefault();
+
+			// If naval yard is too far away, return it.
+			// Else, FindClosest below will find suitable enemy targets :)
+			if (t != null && (t.Location - owner.Units.First().Location).LengthSquared
+					> owner.Bot.Info.MaxBaseRadius * owner.Bot.Info.MaxBaseRadius)
+				return t;
+
+			return owner.Bot.FindClosestEnemy(owner.Units.FirstOrDefault().CenterPosition);
+		}
+	}
+
+	class NavyUnitsIdleState : NavyStateBase, IState
+	{
+		public void Activate(Squad owner) { }
+
+		public void Tick(Squad owner)
+		{
+			if (!owner.IsValid)
+				return;
+
+			if (!owner.IsTargetValid)
+			{
+				var t = FindClosestEnemy(owner);
+				if (t == null) return;
+				owner.TargetActor = t;
+			}
+
+			var enemyUnits = owner.World.FindActorsInCircle(owner.TargetActor.CenterPosition, WDist.FromCells(10))
+				.Where(unit => owner.Bot.Player.Stances[unit.Owner] == Stance.Enemy).ToList();
+
+			if (enemyUnits.Any())
```

Comment:
According to https://github.com/OpenRA/OpenRA/wiki/Coding-Standard, you should return early and reduce the amount of nesting when possible, i.e. make this
```
if (!enemyUnits.Any()
	return;
```